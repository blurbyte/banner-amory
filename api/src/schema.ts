// merged GraphQL schema

import { makeExecutableSchema } from 'apollo-server';
import { merge } from 'lodash';

import { typeDefs as scalarsTypeDefs, resolvers as scalarsResolvers } from './types/scalars';

import { Root } from './types/Root';
import { Item } from './types/Item';
import { Comment } from './types/Comment';

import { rootQueries } from './queries/root';
import { itemQueries } from './queries/item';

import { rootMutations } from './mutations/root';

// merged types
const typeDefs = [scalarsTypeDefs, Root, Item, Comment];

// merged resolvers
const resolvers = merge(
  // queries
  {}, scalarsResolvers, rootQueries, itemQueries,
  // mutations
  rootMutations
);

const schema = makeExecutableSchema({
  typeDefs,
  resolvers
});

export default schema;
