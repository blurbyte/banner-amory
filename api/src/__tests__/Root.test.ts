import { graphql } from 'graphql';
import { addMockFunctionsToSchema } from 'apollo-server';
import { find } from 'lodash';

import schema from '../schema';
import { testItem1, testItem2 } from './mocks';

// arrange
addMockFunctionsToSchema({
  schema,
  mocks: {
    Int: () => null,
    Boolean: () => null,
    String: () => null,
    Query: () => ({
      items: () => [testItem1, testItem2],
      item: (_root, { slug }) => {
        const items = [testItem1, testItem2];
        const item = find(items, item => item.slug === slug);
        return item;
      }
    })
  }
});

const getAllItems = `
  query getAllItems {
    items {
      id
      name
      slug
      gamePart
      description
      statStrength
      aggro
      acquirementMarketplace
    }
  }
`;

const getSingleItem = `
  query getSingleItem($slug: String!) {
    item(slug: $slug) {
      id
      name
      slug
    }
  }
`;

describe('Root queries', () => {
  it('should fetch list of items', async () => {
    // act
    const result = await graphql(schema, getAllItems);

    // assert
    expect(result).toMatchSnapshot();
  });

  it('should fetch single item', async () => {
    // act
    let params = { slug: 'alettes-bracelet' };
    let result = await graphql(schema, getSingleItem, null, null, params);

    // assert
    expect(result).toMatchSnapshot();
  });

  it('should return null if wrong slug got passed', async () => {
    // act
    let params = { slug: 'wrong-one' };
    let result = await graphql(schema, getSingleItem, null, null, params);

    // assert
    expect(result).toMatchSnapshot();
  });
});
