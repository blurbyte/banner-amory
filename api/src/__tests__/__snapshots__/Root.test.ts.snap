// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root queries should get list of items 1`] = `
Object {
  "data": Object {
    "items": Array [
      Object {
        "acquirementMarketplace": null,
        "aggro": null,
        "description": "Lovingly made by a hunter's wife to keep her daughter safe from harm.",
        "gamePart": 1,
        "id": 1,
        "name": "Alette's Bracelet",
        "slug": "alettes-bracelet",
        "statStrength": 1,
      },
      Object {
        "acquirementMarketplace": true,
        "aggro": -1,
        "description": "A former skald's lightweight boots that attracted the wrong type of attention.",
        "gamePart": 2,
        "id": 53,
        "name": "Atfari's Golden Boots",
        "slug": "atfaris-golden-boots",
        "statStrength": null,
      },
    ],
  },
}
`;

exports[`Root queries should get list of items based on search query 1`] = `
Object {
  "data": Object {
    "search": Array [
      Object {
        "id": 1,
        "name": "Alette's Bracelet",
        "slug": "alettes-bracelet",
      },
    ],
  },
}
`;

exports[`Root queries should get single item 1`] = `
Object {
  "data": Object {
    "item": Object {
      "id": 1,
      "name": "Alette's Bracelet",
      "slug": "alettes-bracelet",
    },
  },
}
`;

exports[`Root queries should return empty array if search query doesn't match 1`] = `
Object {
  "data": Object {
    "search": Array [],
  },
}
`;

exports[`Root queries should return null if wrong slug got passed 1`] = `
Object {
  "data": Object {
    "item": null,
  },
}
`;
